src/content/blog/2023-09-26-preparing-custom-rules-eslint-v9.md

```diff
@@ -0,0 +1,245 @@
+        let currentSegments;
+
+        // tracks all current segments for all open paths
+        const allCurrentSegments = [];
+
+        return {
+
+            onCodePathStart(codePath) {
+                currentCodePath = codePath;
+                allCurrentSegments.push(currentSegments);
+                currentSegments = new Set();
+            },
+
+            onCodePathEnd(codePath) {
+                currentCodePath = codePath.upper;
+                currentSegments = allCurrentSegments.pop();
+            },
+
+            onCodePathSegmentStart(segment) {
+                currentSegments.add(segment);
+            },
+
+            onCodePathSegmentEnd(segment) {
+                currentSegments.delete(segment);
+            },
+
+            onUnreachableCodePathSegmentStart(segment) {
+                currentSegments.add(segment);
+            },
+
+            onUnreachableCodePathSegmentEnd(segment) {
+                currentSegments.delete(segment);
+            }
+        };
+
+    }
+};
+```
+
+We have already made this change in all of the ESLint core rules to validate that the approach works as expected.
+
+## `context` methods becoming properties
+
+As we look towards the API we'd like rules for other languages to have, we decided to convert some methods on `context` to properties. The methods in the following table just returned some data that didn't change, so there was no reason they couldn't be properties instead.
+
+|**Deprecated on `context`**|**Property on `context`**|
+|-----------------------|--------------------------|
+|`context.getSourceCode()`|`context.sourceCode`|
+|`context.getFilename()`|`context.filename`|
+|`context.getPhysicalFilename()`|`context.physicalfilename`|
+|`context.getCwd()`|`context.cwd`|
+
+We are deprecating the methods in favor of the properties (added in v8.40.0). These methods will be removed in v10.0.0 (not v9.0.0) as they are not blocking language plugins work. Here's an example that ensures the correct value is used:
+
+```js
+module.exports = {
+    create(context) {
+
+        const sourceCode = context.sourceCode ?? context.getSourceCode();
+        const cwd = context.cwd ?? context.getCwd();
+        const filename = context.filename ?? context.getFilename();
+        const physicalFilename = context.physicalFilename ?? context.getPhysicalFilename();
+
+        return {
+            Program(node) {
+                // do something
+            }
+        }
+    }
+};
+```
+
+## `context` properties being removed
+
+Additionally, the `context.parserOptions` and `context.parserPath` properties are deprecated and will be removed in v9.0.0. There is a new `context.languageOptions` property that allows rules to access similar data as `context.parserOptions`. In general, though, rules should not depend on information either in `context.parserOptions` or `context.languageOptions` to determine how they should behave; this information is primarily for debugging purposes.
```

Comment:
I'm not sure if using `context.languageOptions` should generally be discouraged. We're using it in multiple core rules:

https://github.com/search?q=repo%3Aeslint%2Feslint+context.languageOptions&type=code

`parserOptions.ecmaVersion` was unreliable but `languageOptions.ecmaVersion` as an option of ESLint itself rather than a parser-specific feature fixes those issues.

Reply:
I still feel like it's a bit of a bad practice as things like `ecmaVersion` aren't guaranteed to mean anything, especially with so many people using custom parsers.